src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java

```diff
@@ -21,133 +21,505 @@
-            resetSchedular = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(reset, dynamicResetInterval, dynamicResetInterval, TimeUnit.MILLISECONDS);
-            registerMBean();
+            this.measurement = measurement;
+            this.cachedMeasurement = measurement.measure();
+        }
+
+        @Override
+        public String toString()
+        {
+            return "AnnotatedMeasurement{" +
+                   "millisSinceLastRequest=" + millisSinceLastRequest +
+                   ", millisSinceLastMeasure=" + millisSinceLastMeasure +
+                   ", nextProbeDelayMillis=" + nextProbeDelayMillis +
+                   ", probeFuturePending=" + (probeFuture != null && !probeFuture.isDone()) +
+                   ", measurementClass=" + measurement.getClass().getSimpleName() +
+                   ", cachedMeasurement=" + cachedMeasurement +
+                   '}';
         }
     }
 
     /**
-     * Update configuration from {@link DatabaseDescriptor} and estart the update-scheduler and reset-scheduler tasks
-     * if the configured rates for these tasks have changed.
+     * Allows the subclasses to inject ISnitchMeasurement instances back into this common class so that common
+     * functionality can be handled here
+     * @param initialValue The initial value of the measurement, some implementations may use this, others may not
+     * @return a constructed instance of an ISnitchMeasurement interface
      */
-    public void applyConfigChanges()
+    abstract protected ISnitchMeasurement measurementImpl(long initialValue);
+
+    /**
+     * Records a latency. This MUST be cheap as it is called in the fast path
+     */
+    public void receiveTiming(InetAddressAndPort address, long latencyMicros, LatencyMeasurementType measurementType)
+    {
+        if (measurementType == LatencyMeasurementType.IGNORE)
+           return;
+
+        AnnotatedMeasurement sample = samples.get(address);
+
+        if (sample == null)
+        {
+            AnnotatedMeasurement maybeNewSample = new AnnotatedMeasurement(measurementImpl(latencyMicros));
+            sample = samples.putIfAbsent(address, maybeNewSample);
+            if (sample == null)
+                sample = maybeNewSample;
+        }
+
+        if (measurementType == LatencyMeasurementType.READ && sample.millisSinceLastMeasure.get() > 0)
+            sample.millisSinceLastMeasure.lazySet(0);
+
+        sample.measurement.sample(latencyMicros);
+    }
+
+    @VisibleForTesting
+    protected void reset()
+    {
+        scores.clear();
+        samples.clear();
+    }
+
+    @VisibleForTesting
+    protected void updateScores()
     {
-        if (dynamicUpdateInterval != DatabaseDescriptor.getDynamicUpdateInterval())
+        if (!StorageService.instance.isGossipActive())
+            return;
+
+        if (!registered)
```

Comment:
So after looking into this it boils down to the snitch is loaded at configuration time before StorageService (and implicitly MessagingService), so it shouldn't register at that time. From my testing we were either getting the MessagingService singleton 200ms after the snitch during `updateScores` or we were getting it during StorageService initialization.

I really don't like weak maps in this case, especially since we iterate the subscriber collection frequently. Instead I've refactored so that we use a `CopyOnWriteArraySet` (since we very rarely add or remove latency subscribers but iterate them on basically every request) in `MessagingService` and properly add and remove ourselves from that map. I think this is the cleanest way that maintains the performance requirements but let me know if you disagree.

For now I'm punting on adding anything additional to `ILatencySubscriber`, I think as part of the followup unifying load shedding from the current BackPressure code path and the DES code path makes sense in CASSANDRA-14817